src/main/java/com/aws/greengrass/lifecyclemanager/UpdateSystemPolicyService.java

```diff
@@ -121,16 +120,13 @@ protected void runUpdateActions(String deploymentId) {
      *         false if update actions were already in progress
      */
     public boolean discardPendingUpdateAction(String tag) {
-        if (tag.equals(actionInProgress.get())) {
+        final UpdateAction pendingUpdateAction = pendingActions.remove(tag);
+        if (pendingUpdateAction == null) {
+            // Update action is never added or already in progress.
             return false;
         }
-        final UpdateAction pendingUpdateAction = pendingActions.get(tag);
```

Comment:
Fix - race: A thread can reach this point before the update action is completed (but in progress) resulting in returning an incorrect value for `discardPendingUpdateAction` and sending two post update events (one via `runUpdateActions` and other via `discardPendingUpdateAction`). This is because 1/ `runUpdateActions`removes the action from the map only after it is completed 2/ No locking used while accessing/modifying the map . 

Use `remove` to ensure that the updated state is reflected across threads.